### DISCUSSION TITLE
Multiple Textcat Components in Pipeline

### DISCUSSION THREAD

**Discussion by arc12** [explosion/spaCy#8320] [Help: Coding & Implementations] [answered]
I have been investigating how to include two separate textcat components in a pipeline in the interest of efficiency, each dealing with a separate set of categories. My approach *almost* works, but not completely.

I have the following, which takes 2 previously-trained models and works nicely if I use the Language object while it is still in memory, but it fails if I use `to_disk()` and `spacy_load()`.
```
nlp = spacy.load("model1")
nlp2 = spacy.load("model2")
textcat2 = nlp2.get_pipe("textcat")
nlp.component("textcat_g", func=textcat2)
nlp.add_pipe("textcat_g")
```

The error on `spacy.load()` is:
> ValueError: [E002] Can't find factory for 'textcat_g' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Some hand-crafting the (combined) config JSON is effective, but not really a good solution. The changes I find are needed are:
1. change the following to assert that "textcat" is the factory
> [components.textcat_g]
factory = "textcat_g"
2. add a model specification such as
> [components.textcat_g.model]
@architectures = "spacy.TextCatBOW.v1"
exclusive_classes = true
ngram_size = 1
no_output_layer = false
nO = null

I hope someone can give me some advice, as I have experimented with alternative approaches and failed. Prima facie, my approach seems OK, and it feels like an omission for Spacy to fail to persist the component/factory information. OTOH, my ignorance of correct procedure...

**Comment by adrianeboyd** [comment, marked as answer]
Hi, the right way to do this is with the `source` option, which loads a component from an existing pipeline with its full config and model weights/settings:

```python
nlp1.add_pipe("textcat", name="textcat_g", source=nlp2)
```

The first argument is the name of the component in  `nlp2` and `name=` is what you want to name it in `nlp1`.

The same option is available in a config:

```ini
[components.textcat_g]
source="model2"
component="textcat"
```

From a config you can use `spacy assemble` to create a new pipeline if all training is completed, or `spacy train` to train the sourced component further.

When the config for the new trained/assembled pipeline is saved, the `source=` lines will be replaced with the exact config from the source pipeline, so the final config/pipeline doesn't look any different than one that was trained from scratch.

**Reply by arc12** [reply]
Sweet! Economical and effective.
Thanks for taking the time to explain @adrianeboyd
